common/dynamicconfig/mutable_ephemeral_client.go

```diff
@@ -0,0 +1,171 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamicconfig
+
+import (
+	"strings"
+	"sync"
+)
+
+// MutableEphemeralClient is a dynamicconfig.Client implementation that is
+// safely mutable in the presence of an unbounded number of reads and writes.
+// Writers block other writers but not readers. Readers don't block anything.
+// Mutations are batched to the Update function. Data written to objects of this
+// type is not made durable.
+type MutableEphemeralClient struct {
+	*basicClient
+	mu sync.Mutex
+}
+
+// NewMutableEphemeralClient constructs a new MutableEphemeralClient with an
+// empty internal store.
+func NewMutableEphemeralClient(mutations ...Mutation) *MutableEphemeralClient {
+	c := &MutableEphemeralClient{basicClient: newBasicClient()}
+	c.Update(mutations...)
+	return c
+}
+
+// MutationConstraint expresses a configuration constraint (namespace, queue
+// name, or queue type) that is an optional part of a Mutation.
+type MutationConstraint func(map[string]interface{})
+
+// Mutation expresses a modification to a MutableEphemeralClient's internal
+// value store. Mutation functions are applied with
+// MutableEphemeralClient.Update.
+type Mutation func(configValueMap)
+
+// ForNamespaces builds an exact-match namespace MutationConstraint.
+func ForNamespace(ns string) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[Namespace.String()] = ns
+	}
+}
+
+// ForTaskQueueName builds an exact-match task queue name MutationConstraint.
+func ForTaskQueueName(n string) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[TaskQueueName.String()] = n
+	}
+}
+
+// ForTaskType builds an exact-match task type MutationConstraint.
+func ForTaskType(t int) MutationConstraint {
+	return func(m map[string]interface{}) {
+		m[TaskType.String()] = t
+	}
+}
+
+// Set assigns the supplied value along with optional constraints to the
+// indicated key. Any other additional value (possibly with other constraints)
+// is overwritten.
+func Set(k Key, v interface{}, constraints ...MutationConstraint) Mutation {
+	return func(m configValueMap) {
+		kstr := strings.ToLower(k.String())
+		cv := constrainedValue{Value: v, Constraints: map[string]interface{}{}}
+		for _, c := range constraints {
```

Comment:
`c` -> `constraint` <- will be easier to read

Reply:
`m` -> `config`? or `configValueMap` ?

Reply:
Idiomatic Go prefers short names, particularly in short functions or scopes. The long var names we use contributes to the Java smell. I'm not a zealot about variable naming but in general I'd prefer to move the codebase in the direction of more Go-like rather than less Go-like.

Reply:
`c()` is hard to read
whether go like or not go like is irrelevant, if go like means less readable then we should not follow the go guideline.

Reply:
You're expressing an opinion as though it's a universal fact. I'll leave my own opinion of it and instead note that others, including the authors of Go, directly disagree. Often. For example from the Practice of Programming book

> Programmers are often encouraged to use long variable names regardless of context. That is a mistake: clarity is often achieved through brevity.

Reply:
cf